Add support to Cloudwatch datasource for tlsSkipVerify config variable.

pgarceau opened the pull request:
Currently, cloudwatch datasources must check for valid tls certificates
on the endpoint and don't honor the tlsSkipVerify parameter if it is
added to a provisioning configuation. These changes will read that
configuration and properly configure the aws go sdk to ignore tls
certificate chain validation.

While this doesn't address a specific issue, it may provide a solution for #8443.

pgarceau committed:
Add support to Cloudwatch datasource for tlsSkipVerify config variable.

Currently, cloudwatch datasources must check for valid tls certificates
on the endpoint and don't honor the tlsSkipVerify parameter if it is
added to a provisioning configuation. These changes will read that
configuration and properly configure the aws go sdk to ignore tls
certificate chain validation.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=14628) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=14628) before we can accept your contribution.<br/><hr/>**Peter Garceau** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=14628) it.</sub>

pgarceau committed:
Run go fmt on credentials.go.

torkelo commented:
Could you describe a scenario where this would be needed? Cloudwatch API will always have valid TLS certs, if not something is wrong. 

https://github.com/grafana/grafana/issues/8443 is old before all cloudwatch access was performed from the backend.

torkelo commented:
Not sure this is needed to support https / http proxy, what is needed is to set the Proxy property on the transport: 
https://github.com/grafana/grafana/blob/master/pkg/services/notifications/webhook.go#L28

pgarceau commented:
I can't go into all of the specific details of our network, but the short version is that we required this change because for us amazonaws.com does not return a certificate that matches that domain. This is expected on our network, and isn't something that can be changed. Shutting off validation was the easiest way of getting around the errors that were popping up.

Additionally, the Grafana provisioning docs (http://docs.grafana.org/administration/provisioning/) indicate that all datasources support the tlsSkipVerify configuration parameter, so I assumed that it being left out was an oversight. If it doesn't seem appropriate to merge in this change, I would definitely recommend updating the documentation to indicate that Cloudwatch doesn't support tlsSkipVerify, since it took me a while to figure that out for myself.

torkelo commented on pkg/tsdb/cloudwatch/credentials.go:
```diff
@@ -177,11 +182,27 @@ func (e *CloudWatchExecutor) getAwsConfig(dsInfo *DatasourceInfo) (*aws.Config,
+		cfg := &aws.Config{
+			Region:      aws.String(dsInfo.Region),
+			Credentials: creds,
+			HTTPClient: &http.Client{
```
think this should be changed to use the http client returned here: https://github.com/grafana/grafana/blob/1165d098b0d0ae705955f9d2ea104beea98ca6eb/pkg/models/datasource_cache.go#L28 

It has all the correct settings and transport and is what the other data sources use. Then you can remove this if case here.

pgarceau replied:
This is an excellent idea. Switching over to use this HttpClient makes the modifications much simpler and robust. I've gone ahead and updated my branch to use this suggestion.

pgarceau committed:
Switch to using DataSource.GetHttpClient.

This method of creating the HttpClient object to be used for the
AwsCredentials object is simpler and more robust.

pgarceau committed:
Merge remote-tracking branch 'upstream/master' into tls_verify_cloudwatch

buroa commented:
> I can't go into all of the specific details of our network, but the short version is that we required this change because for us amazonaws.com does not return a certificate that matches that domain. This is expected on our network, and isn't something that can be changed. Shutting off validation was the easiest way of getting around the errors that were popping up.
> 
> Additionally, the Grafana provisioning docs (http://docs.grafana.org/administration/provisioning/) indicate that all datasources support the tlsSkipVerify configuration parameter, so I assumed that it being left out was an oversight. If it doesn't seem appropriate to merge in this change, I would definitely recommend updating the documentation to indicate that Cloudwatch doesn't support tlsSkipVerify, since it took me a while to figure that out for myself.

Yikes. If you are using the Isolated AWS regions -- you should probably turn on TLS verification. If you are using VPC endpoints -- you should make sure your DNS server resolves monitoring.(region).fqdn to your VPC endpoint. Last but not lease, if you are definitely using Isolated regions, be sure you rebuild the Grafana server from source + modifying https://github.com/aws/aws-sdk-go/blob/master/aws/endpoints/defaults.go.

Steve

pgarceau commented:
> 
> 
> > I can't go into all of the specific details of our network, but the short version is that we required this change because for us amazonaws.com does not return a certificate that matches that domain. This is expected on our network, and isn't something that can be changed. Shutting off validation was the easiest way of getting around the errors that were popping up.
> > Additionally, the Grafana provisioning docs (http://docs.grafana.org/administration/provisioning/) indicate that all datasources support the tlsSkipVerify configuration parameter, so I assumed that it being left out was an oversight. If it doesn't seem appropriate to merge in this change, I would definitely recommend updating the documentation to indicate that Cloudwatch doesn't support tlsSkipVerify, since it took me a while to figure that out for myself.
> 
> Yikes. If you are using the Isolated AWS regions -- you should probably turn on TLS verification. If you are using VPC endpoints -- you should make sure your DNS server resolves monitoring.(region).fqdn to your VPC endpoint. Last but not lease, if you are definitely using Isolated regions, be sure you rebuild the Grafana server from source + modifying https://github.com/aws/aws-sdk-go/blob/master/aws/endpoints/defaults.go.
> 
> Steve

The whole situation is definitely ugly, but I really don't have any control over the network itself. We are 100% sure that turning off the validation is safe though, so that's why this is the easiest approach. If it turns out that this change looks useful enough to get merged in, it would mean we could use grafana out of the box and wouldn't have to maintain our own build. If not though, we will likely take the approach you mentioned with an updated aws sdk. I also have a feeling that there are other poor souls out there who would benefit from this change, especially after switching to the method that torkelo recommended.

If this doesn't get merged in though, I would still recommend updating the documentation because it seems to indicate that this feature is already available. I'd be willing to put in a separate pull request for that if it looks to be necessary.

buroa commented:
> > > I can't go into all of the specific details of our network, but the short version is that we required this change because for us amazonaws.com does not return a certificate that matches that domain. This is expected on our network, and isn't something that can be changed. Shutting off validation was the easiest way of getting around the errors that were popping up.
> > > Additionally, the Grafana provisioning docs (http://docs.grafana.org/administration/provisioning/) indicate that all datasources support the tlsSkipVerify configuration parameter, so I assumed that it being left out was an oversight. If it doesn't seem appropriate to merge in this change, I would definitely recommend updating the documentation to indicate that Cloudwatch doesn't support tlsSkipVerify, since it took me a while to figure that out for myself.
> > 
> > 
> > Yikes. If you are using the Isolated AWS regions -- you should probably turn on TLS verification. If you are using VPC endpoints -- you should make sure your DNS server resolves monitoring.(region).fqdn to your VPC endpoint. Last but not lease, if you are definitely using Isolated regions, be sure you rebuild the Grafana server from source + modifying https://github.com/aws/aws-sdk-go/blob/master/aws/endpoints/defaults.go.
> > Steve
> 
> The whole situation is definitely ugly, but I really don't have any control over the network itself. We are 100% sure that turning off the validation is safe though, so that's why this is the easiest approach. If it turns out that this change looks useful enough to get merged in, it would mean we could use grafana out of the box and wouldn't have to maintain our own build. If not though, we will likely take the approach you mentioned with an updated aws sdk. I also have a feeling that there are other poor souls out there who would benefit from this change, especially after switching to the method that torkelo recommended.
> 
> If this doesn't get merged in though, I would still recommend updating the documentation because it seems to indicate that this feature is already available. I'd be willing to put in a separate pull request for that if it looks to be necessary.

Yup. We sync the latest Grafana changes every time a push is made and apply the AWS Go SDK endpoints for our other regions we support. It then builds using docker and these commands.

```mkdir -p /go/src/github.com/grafana/
ln -s /grafana/ /go/src/github.com/grafana/grafana
cd /go/src/github.com/grafana/grafana
/tmp/bootstrap.sh
./scripts/build/download-phantomjs.sh
./scripts/build/build-all.sh
```

Output should be a whole bunch of pretty RPMs ready to be used.

pgarceau committed:
Merge branch 'master' into tls_verify_cloudwatch

pgarceau committed:
Merge branch 'master' into tls_verify_cloudwatch
